dd: reuse buffer for the most common cases

patricksjackson opened the pull request:
This is purely a performance improvement for the case of large blocksizes in dd. This reuses a single buffer instead of reallocating the buffer for each block.

For the uncommon case of dealing with conversions, this still creates a copy of the buffer and so doesn't confer any benefit. That section of the codebase requires some more complex refactoring that is probably better to change separately. See #3293 by @jfinkels 

Some very rough benchmarks piping from `/dev/zero` to `/dev/null` show an improvement of about 3x with large buffer sizes, and minor speedups with smaller buffers. Real world usage will see less benefit since there will usually be more time spent reading/writing the buffers.

Testing can be done with something like:

```
time ./target/release/dd status=progress bs=1G count=10 < /dev/zero > /dev/null
```

| | bs=4k count=1000000 | bs=1G count=10 |
| --- | --- | --- |
| (Reused Buffer) dd | 2.02s | 1.94s |
| Coreutils dd | 2.18s | 6.04s |
| GNU dd | 1.26s | 1.16s |

jfinkels reviewed (approved):
Cool!

anastygnome commented:
you can actually make the buffer non mutable at line 609, which is way better performance wise !

anastygnome commented on src/uu/dd/src/dd.rs:
```diff
@@ -609,14 +617,16 @@ fn read_helper<R: Read>(i: &mut Input<R>, bsize: usize) -> std::io::Result<(Read
+    buf.resize(bsize, BUF_INIT_BYTE);
+
     let mut rstat = match i.cflags.sync {
         Some(ch) => i.fill_blocks(&mut buf, ch)?,
```
We can remove the mut : 
` 
Some(ch) => i.fill_blocks(&mut buf, ch)?,
`

anastygnome commented on src/uu/dd/src/dd.rs:
```diff
@@ -626,10 +636,10 @@ fn read_helper<R: Read>(i: &mut Input<R>, bsize: usize) -> std::io::Result<(Read
 
```
`perform_swab(buf);`

anastygnome commented on src/uu/dd/src/dd.rs:
```diff
@@ -600,7 +604,11 @@ impl Write for Output<io::Stdout> {
-fn read_helper<R: Read>(i: &mut Input<R>, bsize: usize) -> std::io::Result<(ReadStat, Vec<u8>)> {
+fn read_helper<R: Read>(
+    i: &mut Input<R>,
+    mut buf: &mut Vec<u8>,
```
`buf: &mut Vec<u8>,`

anastygnome reviewed (changes requested):
Hey, nice work but you can in fact do even better! 

 (I can't push those changes to your branch sadly :) 
I have tested those locally and clippy now shuts up !
@jfinkels maybe we should integrate before a merge

patricksjackson commented:
Thanks @anastygnome, I went back and fixed the clippy warnings. I haven't done much coding in Rust so my processes aren't quite polished yet :)
 
Here's the updated block using `hyperfine`. Since I already sized these to run for a few seconds, and they largely test memory performance, as expected there isn't much change vs `time`. I added an additional block size in the middle that is roughly the max throughput I can achieve on my machine since it comfortably fits in L3 cache. They can be run with 

```
hyperfine "./target/release/dd bs=4k count=1000000 < /dev/zero > /dev/null"
hyperfine "./target/release/dd bs=1M count=20000 < /dev/zero > /dev/null"
hyperfine "./target/release/dd bs=1G count=10 < /dev/zero > /dev/null"
```

  |  bs=4k count=1000000 | bs=1M count=20000 | bs=1G count=10
-- | -- | -- | --
Coreutils dd | 2.14s | 1.98s | 5.98s
Coreutils dd (Reused buffer) | 1.99s | 1.60s | 1.95s
GNU dd | 1.19s | 1.09s | 1.14s

I've also added a BENCHMARKING document explaining what `dd` characteristics need to be understood for measuring its performance. I may have been a little bit on the wordy side :)

sylvestre commented:
bravo :)
